File: Python3/377. Combination Sum IV.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        nums.sort()
        
        @functools.lru_cache(maxsize = None)
        def helper(remain = target):
            if remain == 0:
                return 1
            
            res = 0
            for n in nums:
                if remain - n >= 0:
                    res += helper(remain - n)
                else:
                    break
            
            return res
        
        return helper()

File: Java/377. Combination Sum IV.java
class Solution {
    private Map<Integer, Integer> memo;
    
    public int combinationSum4(int[] nums, int target) {
        Arrays.sort(nums);
        memo = new HashMap();
        return helper(nums, target);
    }
    
    private int helper(int[] nums, int remain) {
        if (remain == 0) {
            return 1;
        }
        
        if (memo.containsKey(remain)) {
            return memo.get(remain);
        }
        
        int res = 0;
        for (int n: nums) {
            if (remain - n >= 0) {
                res += helper(nums, remain - n);
            } else {
                break;
            }
        }
        
        memo.put(remain, res);
        return res;
    }
}